keel-core/src/main/kotlin/com/netflix/spinnaker/keel/actuation/EnvironmentConstraintRunner.kt

```diff
@@ -0,0 +1,250 @@
+package com.netflix.spinnaker.keel.actuation
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Environment
+import com.netflix.spinnaker.keel.api.anyStateful
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.constraints.ConstraintEvaluator
+import com.netflix.spinnaker.keel.api.constraints.ConstraintStatus.PENDING
+import com.netflix.spinnaker.keel.api.constraints.StatefulConstraintEvaluator
+import com.netflix.spinnaker.keel.core.api.EnvironmentArtifactVetoes
+import com.netflix.spinnaker.keel.core.comparator
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import org.slf4j.LoggerFactory
+import org.springframework.context.ApplicationEventPublisher
+import org.springframework.stereotype.Component
+
+/**
+ * This class is responsible for running constraints and queueing them for approval.
+ * Approval into an environment is the responsibility of the [EnvironmentPromotionChecker].
+ */
+@Component
+class EnvironmentConstraintRunner(
+  private val repository: KeelRepository,
+  private val constraints: List<ConstraintEvaluator<*>>,
+  private val publisher: ApplicationEventPublisher
+) {
+  private val log by lazy { LoggerFactory.getLogger(javaClass) }
+
+  private val implicitConstraints: List<ConstraintEvaluator<*>> = constraints.filter { it.isImplicit() }
+  private val explicitConstraints: List<ConstraintEvaluator<*>> = constraints - implicitConstraints
+
+  // constraints that are only run if they are defined in a delivery config
+  private val statefulEvaluators: List<ConstraintEvaluator<*>> = explicitConstraints
+    .filterIsInstance<StatefulConstraintEvaluator<*, *>>()
+  private val statelessEvaluators = explicitConstraints - statefulEvaluators
+
+  // constraints that run for every environment in a delivery config but aren't shown to the user.
+  private val implicitStatefulEvaluators: List<ConstraintEvaluator<*>> = implicitConstraints
+    .filterIsInstance<StatefulConstraintEvaluator<*, *>>()
+  private val implicitStatelessEvaluators: List<ConstraintEvaluator<*>> = implicitConstraints - implicitStatefulEvaluators
+
+  /**
+   * Checks the environment and determines the version that should be approved,
+   * or null if there is no version that passes the constraints for an env + artifact combo.
+   * Queues that version for approval if it exists.
+   */
+  fun checkEnvironment(
+    envContext: EnvironmentContext
+  ) {
+    val pendingVersionsToCheck: MutableSet<String> =
+      when (envContext.environment.constraints.anyStateful) {
+        true -> repository
+          .pendingConstraintVersionsFor(envContext.deliveryConfig.name, envContext.environment.name)
+          .filter { envContext.versions.contains(it) }
+          .toMutableSet()
+        false -> mutableSetOf()
+      }
+
+    checkConstraints(
+      envContext,
+      pendingVersionsToCheck
```

Comment:
`pendingVersionsToCheck` could be at this point either filled with versions that have a pending stateful constraint or null (and then we will check all versions for the artifact, as it probably only have some stateless constraints if any?)

Reply:
yes, `pendingVersionsToCheck` is filled with all pending stateful constraints (i.e. a manual judgement that wasn't judged when this ran last. It could be empty.

we check the versions and find the one we will queue for approval, then go back and take a look at any pending ones we haven't looked at yet in case they were judged or fail because of a timeout.